src/mir/construction.md

```diff
@@ -1 +1,111 @@
+list of items:
+
+* Function and Closure bodies
+* Initializers of `static` and `const` items
+* Initializers of enum discriminants
+* Glue and Shims of any kind
+    * Tuple struct initializer functions
+    * Drop code (the `Drop::drop` function is not called directly)
+    * Drop implementations of types without an explicit `Drop` implementation
+
+The lowering is triggered by calling the `mir_built` query. The entire lowering
+code lives in `src/librustc_mir/build`. There is an intermediate representation
+between [HIR] and [MIR] called the `HAIR` that is only used during the lowering.
+The `HAIR` has datatypes that mirror the [HIR] datatypes, but instead of e.g. `-x`
+being a `hair::ExprKind::Neg(hair::Expr)` it is a `hair::ExprKind::Neg(hir::Expr)`.
+
+This shallowness enables the `HAIR` to represent all datatypes that [HIR] has, but
+without having to create an in-memory copy of the entire [HIR]. The `HAIR` also
+does a few simplifications, e.g. method calls and function calls have been merged
+into a single variant.
+
+The lowering creates local variables for every argument as specified in the signature.
+Next it creates local variables for every binding specified (e.g. `(a, b): (i32, String)`)
+produces 3 bindings, one for the argument, and two for the bindings. Next it generates
+field accesses that read the fields from the argument and writes the value to the binding
+variable.
+
+With this initialization out of the way, the lowering triggers a recursive call
+to a function that generates the MIR for the body (a `Block` expression) and
+writes the result into the `RETURN_PLACE`.
+
+## `unpack!` all the things
+
+One important thing of note is the `unpack!` macro, which accompanies all recursive
+calls. The macro ensures, that you get the result of the recursive call while updating
+the basic block that you are now in. As an example: lowering `a + b` will need to do
+three somewhat independent things:
+
+* give you an `Rvalue` referring to the result of the operation
+* insert an assertion ensuring that the operation does not overflow
+* tell you in which basic block you should write further operations into, because
+  the basic block has changed due to the inserted assertion (assertions terminate
+  blocks and jump either to a panic block or a newly created block, the latter being
+  the one you get back).
+
+The `unpack!` macro will call the recursive function you pass it, return the `Rvalue`
+and update the basic block by mutating the basic block variable you pass to it.
+
+## Lowering expressions into the desired MIR
+
+There are essentially four kinds of representations one might want of a value:
+
+* `Place` refers to a (or a part of) preexisting memory location (local, static, promoted)
+* `Rvalue` is something that can be assigned to a `Place`
+* `Operand` is an argument to e.g. a `+` operation or a function call
+* a temporary variable containing a copy of the value
```

Comment:
It'd be great if we can link to some of the code here. The "handover" between those different bits of code is really sorta confusing and worth trying to document but I wont' force you to do it in this PR =) 

(e.g., it'd be nice to document the "categorization" stuff etc)

Reply:
I addressed everything but this.

> The "handover" between those different bits of code is really sorta confusing and worth trying to document

I discussed this a little bit below in the form of an abstract example. I could add some graphics:

![
digraph G {
    Operand -> Constant
    Operand -> Place
    Place -> Projection
    Projection -> Place
    Place -> Local
    Rvalue -> Operand
    Assignment -> Place
    Assignment -> Operand
}
](https://user-images.githubusercontent.com/332036/47997290-98d04800-e0fb-11e8-9843-ce637b0230a8.png)


Or more verbose

![digraph G {
    Operand -> Constant
    Operand -> Place [taillabel="read"]
    Place -> Projection
    Projection -> Place
    Place -> LocalId
    Rvalue -> "Rvalue\nAggregate"
    Rvalue -> "Rvalue\nBinaryOp"
    Rvalue -> "Rvalue\nUnaryOp"
    Rvalue -> "Rvalue\n..."
    "Rvalue\nAggregate" -> Operand [headlabel="*"]
    "Rvalue\nBinaryOp" -> Operand [headlabel="2"]
    "Rvalue\nUnaryOp" -> Operand
    "Statement\nAssignment" -> Place [taillabel="write"]
    "Statement\nAssignment" -> Rvalue
    Statement -> "Statement\nAssignment"
    Statement -> "Statement\n..."
    Block -> Statement [headlabel="*"]
    Block -> Terminator
    Terminator -> "Terminator\nSwitchInt"
    "Terminator\nSwitchInt" -> Operand
    "Terminator\nSwitchInt" -> Constant [headlabel="*"]
    "Terminator\nSwitchInt" -> BlockId [headlabel="*"]
    Terminator -> "Terminator\n..."
    Mir -> Block [headlabel="*"]
    Mir -> Local [headlabel="*"]
}](https://user-images.githubusercontent.com/332036/47997756-42640900-e0fd-11e8-8446-957fa8c7ceef.png)

Reply:
I wrote them in dot format (the dot source is in the comment source, which I guess is only visible with edit powers to the comment) and used graphviz (or specifically https://dreampuf.github.io/GraphvizOnline/) to generate the pngs. I'd prefer svgs (which can also be generated), but I can't include them in github comments.

When including them in this repo I'd like to include the .dot and .svg files and make the html page point to the svgs. The dot files are important for being able to edit the images (by regenerating them from scratch after editing the dot source)

I think the big one should probably not be inline, but rather mentioned like 
[Click here for a more detailed graph](https://user-images.githubusercontent.com/332036/47997290-98d04800-e0fb-11e8-9843-ce637b0230a8.png)

Reply:
I mean if we feel fancy we can also generate such graphs with rustdoc directly when documenting crates, but it's most likely going to be less readable than the above, due to everything being connected to everything in some way.